src/gas_metering/mod.rs

```diff
@@ -711,22 +723,47 @@ fn generate_gas_counter(gas_global: u32) -> (Type, Function) {
 fn inject_counter<R: Rules>(
     instructions: &wasmparser::FunctionBody,
     rules: &R,
+    param_count: u32,
     gas_func: u32,
 ) -> Result<wasm_encoder::Function> {
-    let blocks = determine_metered_blocks(instructions, rules)?;
-    insert_metering_calls(instructions, blocks, gas_func)
+    let (blocks, metered_instrs) = determine_metered_blocks(instructions, rules)?;
+    insert_metering_calls(instructions, blocks, metered_instrs, param_count, gas_func)
 }
 
 // Then insert metering calls into a sequence of instructions given the block locations and costs.
 fn insert_metering_calls(
     func_body: &wasmparser::FunctionBody,
     blocks: Vec<MeteredBlock>,
+    instructions: Vec<MeteredInstruction>,
+    param_count: u32,
     gas_func: u32,
 ) -> Result<wasm_encoder::Function> {
-    let mut new_func = wasm_encoder::Function::new(copy_locals(func_body)?);
+    // collect value types on which we will be doing dynamic gas math.
+    // We need those for temp locals because wasm has no other way to duplicate stack items..
```

Comment:
Gah, I forgot how much I hate this.

FYI, we could use a global. But I'd really rather not (probably has terrible perf).

Reply:
I thought about that, but it wasn't clear if wasmtime is able to optimize data dependencies through globals. It does resolve through locals, so the cost calc in theory should get JIT-ed into a single register + 3/4 x86 instructions